src/HIE/Bios/Ghc/Gap.hs

```diff
@@ -131,25 +291,234 @@ outType :: MatchGroup GhcTc LExpression -> Type
+#elif __GLASGOW_HASKELL__ >= 808
+numLoadedPlugins = length . Plugins.plugins . hsc_dflags
 #else
 -- Plugins are loaded just as they are used
 numLoadedPlugins _ = 0
 #endif
 
-initializePlugins :: HscEnv -> DynFlags -> IO DynFlags
-#if __GLASGOW_HASKELL__ >= 808
-initializePlugins = DynamicLoading.initializePlugins
+initializePluginsForModSummary :: HscEnv -> ModSummary -> IO (Int, [G.ModuleName], ModSummary)
+initializePluginsForModSummary hsc_env' mod_summary = do
+#if __GLASGOW_HASKELL__ >= 902
+  hsc_env <- DynamicLoading.initializePlugins hsc_env'
+  pure ( numLoadedPlugins hsc_env
+       , pluginModNames $ hsc_dflags hsc_env
+       , mod_summary
+       )
+#elif __GLASGOW_HASKELL__ >= 808
+  let dynFlags' = G.ms_hspp_opts mod_summary
+  dynFlags <- DynamicLoading.initializePlugins hsc_env' dynFlags'
+  pure ( numLoadedPlugins $ set_hsc_dflags dynFlags hsc_env'
+       , G.pluginModNames dynFlags
+       , mod_summary { G.ms_hspp_opts = dynFlags }
+       )
 #else
--- In earlier versions of GHC plugins are just loaded before they are used.
-initializePlugins _ df = return df
+  -- In earlier versions of GHC plugins are just loaded before they are used.
+  return (numLoadedPlugins hsc_env', G.pluginModNames $ hsc_dflags hsc_env', mod_summary)
 #endif
 
-unsetLogAction :: GhcMonad m => m ()
-unsetLogAction =
-    setLogAction (\_df _wr _s _ss _pp _m -> return ())
-#if __GLASGOW_HASKELL__ < 806
-        (\_df -> return ())
+
+setFrontEndHooks :: Maybe (ModSummary -> G.Hsc Tc.FrontendResult) -> HscEnv -> HscEnv
+setFrontEndHooks frontendHook env =
+#if __GLASGOW_HASKELL__ >= 902
+  env
+    { hsc_hooks = hooks
+        { hscFrontendHook = frontendHook
+        }
+    }
+  where
+    hooks = hsc_hooks env
+#else
+  env
+    { G.hsc_dflags = flags
+        { G.hooks = oldhooks
+            { hscFrontendHook = frontendHook
+            }
+        }
+    }
+  where
+    flags = hsc_dflags env
+    oldhooks = G.hooks flags
+#endif
+
+#if __GLASGOW_HASKELL__ < 902
+type Logger = ()
+#endif
+
+getLogger :: HscEnv -> Logger
+getLogger =
+#if __GLASGOW_HASKELL__ >= 902
+    hsc_logger
+#else
+    const ()
+#endif
+
+gopt_set :: DynFlags -> G.GeneralFlag -> DynFlags
+gopt_set = G.gopt_set
+
+setWayDynamicIfHostIsDynamic :: DynFlags -> DynFlags
+setWayDynamicIfHostIsDynamic =
+  if hostIsDynamic
+    then
+      updateWays . addWay' WayDyn
+    else
+      id
+
+#if __GLASGOW_HASKELL__ >= 900
+updateWays :: DynFlags -> DynFlags
+updateWays = id
+
+#if __GLASGOW_HASKELL__ >= 902
+-- Copied from GHC, do we need that?
+addWay' :: Way -> DynFlags -> DynFlags
+addWay' w dflags0 =
+   let platform = targetPlatform dflags0
+       dflags1 = dflags0 { targetWays_ = Platform.addWay w (targetWays_ dflags0) }
+       dflags2 = foldr setGeneralFlag' dflags1
+                       (Platform.wayGeneralFlags platform w)
+       dflags3 = foldr unSetGeneralFlag' dflags2
+                       (Platform.wayUnsetGeneralFlags platform w)
+   in dflags3
+#endif
```

Comment:
IIRC, we don't need this in later GHC versions, as we can use `GHC.Driver.Session.setDynamicNow` directly.